six/modules/c++/six.sicd/source/Grid.cpp

```diff
@@ -62,6 +71,384 @@ bool DirectionParameters::operator==(const DirectionParameters& rhs) const
         weights == rhs.weights);
 }
 
+std::pair<double, double> DirectionParameters::calculateDeltaKs(
+        const ImageData& imageData) const
+{
+    //Here, we assume the min and max of DeltaKCOAPoly must be
+    // on the vertices of the image, since it is smooth and monotonic in most cases--
+    // although in actuality this is not always the case.  To be totally generic, 
+    // we would have to search for an interior min and max as well
+    double derivedDeltaK1 = 0;
+    double derivedDeltaK2 = 0;
+
+    std::vector<RowColInt> vertices = calculateImageVertices(imageData);
+
+    if (!Init::isUndefined<Poly2D>(deltaKCOAPoly))
+    {
+        derivedDeltaK1 = std::numeric_limits<double>::infinity();
+        derivedDeltaK2 = -std::numeric_limits<double>::infinity();
+
+        for (size_t ii = 0; ii < vertices.size(); ++ii)
+        {
+            double currentDeltaK = deltaKCOAPoly.atY(
+                    static_cast<double>(vertices[ii].col))(
+                    static_cast<double>(vertices[ii].row));
+            derivedDeltaK1 = std::min(currentDeltaK, derivedDeltaK1);
+            derivedDeltaK2 = std::max(currentDeltaK, derivedDeltaK2);
+        }
+    }
+
+    derivedDeltaK1 -= (impulseResponseBandwidth / 2);
+    derivedDeltaK2 += (impulseResponseBandwidth / 2);
+
+    if (derivedDeltaK1 < -(1 / sampleSpacing) / 2 ||
+            derivedDeltaK2 > (1 / sampleSpacing) / 2)
+    {
+        derivedDeltaK1 = -(1 / sampleSpacing) / 2;
+        derivedDeltaK2 = -derivedDeltaK1;
+    }
+
+    return std::pair<double, double>(derivedDeltaK1, derivedDeltaK2);
+}
+
+std::auto_ptr<Functor>
+DirectionParameters::calculateWeightFunction() const
+{
+    std::auto_ptr<Functor> weightFunction;
+
+    if (weightType.get() != NULL)
+    {
+        std::string windowName(weightType->windowName);
+        str::upper(windowName);
+
+        if (windowName == "UNIFORM")
+        {
+            weightFunction.reset(new Identity());
+        }
+        else if (windowName == "HAMMING")
+        {
+            double coef;
+            if (weightType->parameters.empty() || weightType->parameters[0].str().empty())
+            {
+                //A Hamming window is defined in many places as a raised cosine of weight .54,
+                //so this is the default. However, some data use a generalized raised cosine and
+                //call it HAMMING, so we allow for both uses.
+                coef = .54;
+            }
+            else
+            {
+                coef = str::toType<double>(weightType->parameters[0].str());

```

Comment:
In this and at least one other similar spot below, doesn't it work to just do...
`coef = weightType->parameters[0];`?
The `operator double()` (which I don't necessarily love but it's intended for stuff like this) should just do this for you...